### PR TITLE
packagekit: Drop usage of D-Bus proxies

### PR DESCRIPTION
**Pull request by martinpitt**
In most cases we don't require object properties or always subscribing to their changes. We also never removed signal listeners from finished transactions. The excessive D-Bus calls from the proxy seem to upset PackageKit sometimes, resulting in failed calls to `GetUpdates()` or `GetOldTransactions()`, such as `TypeError: undefined is not a function`. (Observed e. g. [here](https://fedorapeople.org/groups/cockpit/logs/pull-7169-20170706-235757-40d51050-verify-ubuntu-stable/log.html#82))
    
Restructure `pkTransaction()` to accept a signal handler map and a property change handler, and do the plumbing right in that function, as it's the same for all kinds of PackageKit transactions. This allows us to change/fix the transaction logic in one place.

Also fix a race condition in the tests.

### CODE REVIEW

File: pkg/packagekit/updates.jsx [hunk 1]
```diff
@@ -56,18 +56,43 @@ const PK_STATUS_STRINGS = {
     14: _("Verifying"),
 }
 
+const I_TRANS = "org.freedesktop.PackageKit.Transaction";
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: Since this is used pretty rarely, I wonder if we shouldn't name this variable more descriptively. Also, why should it be in caps?

File: pkg/packagekit/updates.jsx [hunk 2]
```diff
@@ -56,18 +56,43 @@ const PK_STATUS_STRINGS = {
...
+
 var dbus_pk = cockpit.dbus("org.freedesktop.PackageKit", {superuser: "try", "track": true});
 var packageSummaries = {};
 
-function pkTransaction() {
+function pkWatchTransaction(transObj, signalHandlers, notifyHandler) {
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: Do you think `transactionObject` would be too long here? The other variables aren't abbreviated.

File: pkg/packagekit/updates.jsx [hunk 3]
```diff
@@ -56,18 +56,43 @@ const PK_STATUS_STRINGS = {
...
+    var subscriptions = [];
+
+    for (let h in signalHandlers)
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: Should we call this `handler`? It isn't used often.
Also the for loop might look better with braces `{ }`. I'm not entirely sure about this, since the loop only contains one statement, yet that is spread over two lines...

File: pkg/packagekit/updates.jsx [hunk 4]
```diff
@@ -226,15 +251,25 @@ class ApplyUpdates extends React.Component {
...
-            // info: see PK_STATUS_* at https://github.com/hughsie/PackageKit/blob/master/lib/packagekit-glib2/pk-enum.h
-            this.setState({curPackage: pfields[0] + " " + pfields[1],
-                           curStatus: info,
-                           percentage: transProxy.Percentage <= 100 ? transProxy.Percentage : 0,
-                           timeRemaining: transProxy.RemainingTime > 0 ? transProxy.RemainingTime : null});
+        var transObj = this.props.transaction;
+
+        pkWatchTransaction(transObj, {
+            Package: (info, packageId) => {
+                var pfields = packageId.split(";");
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: I believe using `let` here would be more consistent with the rest of the code.

File: pkg/packagekit/updates.jsx [hunk 5]
```diff
@@ -226,15 +251,25 @@ class ApplyUpdates extends React.Component {
...
+
+                dbus_pk.call(transObj, "org.freedesktop.DBus.Properties", "GetAll", [I_TRANS], {timeout: 500})
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: Why is the timeout significantly lower here (by an order of magnitude) than with the other calls? Should the timeout be consistent and maybe in a variable?

File: pkg/packagekit/updates.jsx [hunk 6]
```diff
@@ -369,76 +400,72 @@ class OsUpdates extends React.Component {
...
-                });
+                },
+
+            },  // end pkTransaction signalHandlers
+
+            notify => {
+                if ("Status" in notify) {
+                    let waiting = (notify.Status == PK_STATUS_ENUM_WAIT || notify.Status == PK_STATUS_ENUM_WAITING_FOR_LOCK);
+                    if (waiting != this.state.waiting) {
+                        // to avoid flicker, we only switch to "locked" after 1s, as we will get a WAIT state
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: `after this` maybe?

File: pkg/packagekit/updates.jsx [hunk 7]
```diff
@@ -369,76 +400,72 @@ class OsUpdates extends React.Component {
...
+                        // even if the package db is unlocked
+                        if (waiting) {
+                            this.setState({waiting: true});
+                            window.setTimeout(() => {!this.state.waiting || this.setState({state: "locked"})}, 1000);
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: This timeout is somewhere in the middle... on purpose? The others were `500` and `5000` ms.

File: pkg/packagekit/updates.jsx [hunk 8]
```diff
@@ -457,56 +484,55 @@ class OsUpdates extends React.Component {
...
-            }
-        });
+                Finished: exit => {
+                    this.setState({applyTransaction: null, allowCancel: null});
 
-        transProxy.addEventListener("changed", (event, data) => {
-            if ("AllowCancel" in data)
-                this.setState({allowCancel: data.AllowCancel});
-        });
+                    if (exit == PK_EXIT_ENUM_SUCCESS)
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: Since the other parts of this `if` have braces, so should this one.

Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: If this is a number, wouldn't `===` be more suited to use as the comparison operator?

File: pkg/packagekit/updates.jsx [hunk 9]
```diff
@@ -457,56 +484,55 @@ class OsUpdates extends React.Component {
...
+                        this.setState({state: "updateSuccess", haveSecurity: false, loadPercent: null});
+                    else if (exit == PK_EXIT_ENUM_CANCELLED) {
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: If this is a number, wouldn't `===` be more suited to use as the comparison operator?

File: pkg/packagekit/updates.jsx [hunk 10]
```diff
@@ -457,56 +484,55 @@ class OsUpdates extends React.Component {
...
+                        this.setState({state: "loading", haveSecurity: false, loadPercent: null});
+                        this.loadUpdates();
+                    } else {
+                        // normally we get FAILED here with ErrorCodes; handle unexpected errors to allow for some debugging
+                        if (exit != PK_EXIT_ENUM_FAILED)
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: If this is a number, wouldn't `!==` be more suited to use as the comparison operator?

File: pkg/packagekit/updates.jsx [hunk 11]
```diff
@@ -457,56 +484,55 @@ class OsUpdates extends React.Component {
...
-                              this.setState({state: "updateError"});
-                          });
+        var ids = Object.keys(this.state.updates);
+        if (securityOnly)
+            ids = ids.filter(id => this.state.updates[id].security);
+
+        pkTransaction("UpdatePackages", [0, ids], {}, null, ex => {
+                // We get more useful error messages through ErrorCode or "PackageKit has crashed", so only
+                // show this if we don't have anything else
+                if (this.state.errorMessages.length == 0)
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: `===`

File: pkg/packagekit/updates.jsx [hunk 12]
```diff
@@ -457,56 +484,55 @@ class OsUpdates extends React.Component {
...
+                    this.setState({applyTransaction: null, allowCancel: null});
 
-        transProxy.addEventListener("changed", (event, data) => {
-            if ("AllowCancel" in data)
-                this.setState({allowCancel: data.AllowCancel});
-        });
+                    if (exit == PK_EXIT_ENUM_SUCCESS)
+                        this.setState({state: "updateSuccess", haveSecurity: false, loadPercent: null});
+                    else if (exit == PK_EXIT_ENUM_CANCELLED) {
+                        this.setState({state: "loading", haveSecurity: false, loadPercent: null});
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: I'm unsure about why `haveSecurity` has to be reset here and above, but the code was already present before this pull request...

File: pkg/packagekit/updates.jsx [hunk 13]
```diff
@@ -638,7 +662,8 @@ class OsUpdates extends React.Component {
             <div>
                 <HeaderBar state={this.state.state} updates={this.state.updates}
                            timeSinceRefresh={this.state.timeSinceRefresh} onRefresh={this.handleRefresh}
-                           allowCancel={this.state.allowCancel} onCancel={() => this.state.applyTransaction.Cancel()} />
+                           allowCancel={this.state.allowCancel}
+                           onCancel={() => dbus_pk.call(this.state.applyTransaction, I_TRANS, "Cancel", [])} />
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: I think we usually add spaces inside the braces for more complex statements `{ () => func }`

File: pkg/packagekit/updates.jsx [hunk 14]
```diff
@@ -598,27 +624,25 @@ class OsUpdates extends React.Component {
...
-                });
+                },
+            },
 
-                transProxy.addEventListener("changed", (event, data) => {
-                    if ("Percentage" in data && data.Percentage <= 100)
-                        this.setState({loadPercent: data.Percentage});
-                });
+            notify => {
+                if ("Percentage" in notify && notify.Percentage <= 100)
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: [Above](https://github.com/cockpit-project/cockpit/pull/7175/files#diff-6d5babc8cfb37a2e581d9cd085209d90R262) you changed the usage to `Percentage.v`, is this different?

File: pkg/packagekit/updates.jsx (hunk 1 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: caps> old habit for naming constants from C programming. The various `PK_{EXIT,INFO}_*` constants are also all caps as they are defined like that in the [C headers](https://github.com/hughsie/PackageKit/blob/master/lib/packagekit-glib2/pk-enum.h).

I use it about 10 times, and wanted to keep it shourt. But happy to rename to `transactionInterface` - it seems there is no real convention in JS for naming constants. Fixed locally (will push once I addressed everything else).

File: pkg/packagekit/updates.jsx (hunk 2 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: Renamed to `transactionPath`, as it's only a string (object path), not really an object.

File: pkg/packagekit/updates.jsx (hunk 3 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: Done.

File: pkg/packagekit/updates.jsx (hunk 4 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: Indeed, well spotted. I also found two similar places, fixed.

File: pkg/packagekit/updates.jsx (hunk 5 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: For starting transactions a long timeout (5s) is fine. But this is a handler for PackageKit's "progress" signals, which occur very often (magnitude of once per second). These handlers shouldn't overlap (too much), this is why I set a low timeout here. It's better to miss one call (just in case) than eternally limping behind. I added a comment to explain.

File: pkg/packagekit/updates.jsx (hunk 6 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: Sorry, what do you mean exactly?

File: pkg/packagekit/updates.jsx (hunk 7 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: That's the "1s" from the comment above. This is the timeout after which we switch state from "Loading updates..." to "Some other program is currently using the package manager.." if we get a signal that the DB is locked. One second seems long enough to not appear as "flicker" any more, but still short enough to be reactive to what PackageKit says. This is of course utterly subjective, so I'm fine with changing that to 5s as well.

The 5s are a technical timeout, while this is a visual one.

File: pkg/packagekit/updates.jsx (hunk 8 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: Braces added.

I need to teach myself to be aware of `===`. I changed all `PK_*` constant comparisons accordingly, but I do a separate commit for that as that's unrelated to the proxysectomy.

File: pkg/packagekit/updates.jsx (hunk 9 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: Fixed in the same commit as for the previous  comment.

File: pkg/packagekit/updates.jsx (hunk 12 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: Indeed it's unnecessary here, as the update list should be the same as before when cancelling. I dropped it.

File: pkg/packagekit/updates.jsx (hunk 14 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: Yes, it is. Apparently the `dbus_client.watch()` `notify` callback sends "unwrapped" objects (console.log shows `handleRefresh notify Object { Percentage: 64 }`) while the raw `DBus.Properties.GetAll()` returns the JSON representation of GVariants (something like `Percentage": {"t": "u", "v": "63"}`).

File: pkg/packagekit/updates.jsx (hunk 13 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: Ack. Done as a separate commit as it's rather intrusive and unrelated to the proxysectomy.

File: pkg/packagekit/updates.jsx (hunk 5 above)
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: Perfect, thanks

File: pkg/packagekit/updates.jsx (hunk 6 above)
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: reading fail... I read `is` instead of `1s`...

File: pkg/packagekit/updates.jsx (hunk 7 above)
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: I misread the comment above, so this makes sense

File: pkg/packagekit/updates.jsx (hunk 14 above)
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: yay for consistency :S

File: pkg/packagekit/updates.jsx [hunk 15]
```diff
@@ -457,56 +487,55 @@ class OsUpdates extends React.Component {
...
-                this.setState({allowCancel: data.AllowCancel});
-        });
+                    if (exit === PK_EXIT_ENUM_SUCCESS) {
+                        this.setState({ state: "updateSuccess", haveSecurity: false, loadPercent: null });
+                    } else if (exit === PK_EXIT_ENUM_CANCELLED) {
+                        this.setState({ state: "loading", loadPercent: null });
+                        this.loadUpdates();
+                    } else {
+                        // normally we get FAILED here with ErrorCodes; handle unexpected errors to allow for some debugging
+                        if (exit != PK_EXIT_ENUM_FAILED)
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: this was forgotten, but not important enough to change the commit again

Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: I changed that (and a similar spot) in #7169.
